File: lib/email_inquire/validator/unique_domain_provider.rb
# frozen_string_literal: true

require "email_inquire/validator/base"

module EmailInquire
  module Validator
    class UniqueDomainProvider < Base

      DOMAINS = load_data("unique_domain_providers").freeze

      def validate
        return response.valid! if DOMAINS.include?(domain)

        base, _tld = domain.split(".", 2)

        replacement_domain =
          DOMAINS.find do |reference|
            reference_base, _reference_tld = reference.split(".")

            reference_base.eql?(base)
          end

        response.hint!(domain: replacement_domain) if replacement_domain
      end

    end
  end
end

File: implementations/dotnet/EmailInquire/Validators/UniqueDomainProvider.cs
using System.Linq;

namespace EmailInquire.Validators
{
    public class UniqueDomainProvider : ValidatorBase<UniqueDomainProvider>
    {
        public override Response Validate()
        {
            if (UniqueDomainProviders.Value.Contains(Domain)) return Response.Valid(Email);
            var domain = UniqueDomainProviders.Value.FirstOrDefault(d => d.Split('.')[0].Equals(Domain.Split('.')[0]));
            return !string.IsNullOrEmpty(domain) ? Response.Hint(Name, domain) : Response.Undefined;
        }
    }
}